### ISSUE TITLE
Consider a PyTorch-style shared libcytnx.so to restore C++-extension support without the static-lib bloat

### ISSUE THREAD

**Issue by IvanaGyro** [Cytnx-dev/Cytnx#966]
## Context

#947 found that `libcytnx.a` (the full compiled C++ backend, duplicated as a static archive) is the single largest contributor to oversized PyPI wheels — ~64 MB packed on manylinux x86_64, ~70% of that wheel. The fix landed there (excluding `ARCHIVE`/headers/CMake config from the Python wheel's `install.components`, see #967) removes that bloat, but as a side effect it also drops `cytnx.__cpp_include__`/`cytnx.__cpp_lib__` and the C++-interop workflow documented in `docs/source/install.rst`/`adv_install.rst`, since those attributes pointed at the now-unshipped headers/static lib.

That workflow had real precedent — see the comparison research in #947: numpy ships two narrow static-lib utility slices (`libnpymath.a`/`libnpyrandom.a`, ~50-70 KB each) alongside `numpy.get_include()`, and PyTorch ships a full **shared**-library copy of its compiled core (`libtorch.so`/`libtorch_cpu.so`/`libc10.so`, confirmed to export real linkable C++ symbols) alongside `torch.utils.cpp_extension.include_paths()`/`library_paths()`. cytnx's static-lib approach put it in the same camp as those two — the problem was never "shipping something for downstream C++ builds," it was that cytnx's linkable artifact was (a) the *entire* backend rather than a narrow slice, and (b) a raw, LTO-bloated static archive (see #947's GIMPLE-IR finding) rather than a properly-linked shared library.

This issue tracks the alternative to simply dropping the feature: restructure `cytnx` as a shared library, PyTorch-style, so the C++-interop workflow can come back without reintroducing the size/bloat problem.

**Related PR:** #967 implements the "drop it for now" side of this tradeoff (excludes static lib/headers from the wheel, removes `__cpp_include__`/`__cpp_lib__`, updates docs). This issue is the follow-up for restoring the feature properly instead of leaving it dropped.

## What changes this would require

Traced against the current build (`CMakeLists.txt`/`CytnxBKNDCMakeLists.cmake`):

**The one large item:**
- `cytnx` currently sets `C_VISIBILITY_PRESET hidden` + `VISIBILITY_INLINES_HIDDEN ON` (`CMakeLists.txt:234-235`), and there is **zero** existing export-macro infrastructure anywhere in the tree (no `CYTNX_API`, no `__attribute__((visibility(...)))`, no `dllexport`). A shared `libcytnx.so` today would export almost nothing. PyTorch solves this with `TORCH_API`/`C10_API` macros on every publicly-used class/function. cytnx would need the equivalent across all **55 public headers** under `include/`. This is the dominant cost of the whole change — large, mechanical, but not conceptually hard.

**Smaller, mechanical CMake/build changes:**
- `CMakeLists.txt:233` — `add_library(cytnx STATIC)` → `SHARED`.
- `CMakeLists.txt:416` — `pycytnx`'s link to `cytnx` needs an RPATH to the now-separate `libcytnx.so` (pycytnx already has platform-conditional RPATH logic for its vendor deps at lines ~430-440; the same treatment is needed for `libcytnx.so`, plus deciding where it installs relative to the `cytnx/` package).
- `INSTALL(TARGETS cytnx ...)`'s component split changes: no more `.a`; the new `LIBRARY` output is a genuine runtime dependency, so it needs `COMPONENT libraries` (not `Development`) and comes back into the Python wheel deliberately.
- `CytnxBKNDCMakeLists.cmake` — the `hptt_static`-bundled-into-`cytnx_targets`-export logic (with its explanatory comment about why a STATIC `cytnx` can't absorb another archive) goes away: a shared `cytnx` absorbs `hptt_static`'s objects directly, so `hptt_static` no longer needs its own export-set entry.
- `cytnx/__init__.py` — `_resolve_cpp_linkflags__()`'s computed flags change shape (`-lcytnx` + rpath instead of positional `.a` paths); `__cpp_include__`/`__cpp_lib__` would need reintroducing pointing at the new shared-lib layout.
- `docs/source/install.rst`/`adv_install.rst` — re-document the C++-interop workflow against the new shared-lib shape.
- `cmake/CytnxConfig.cmake.in` / `install(EXPORT cytnx_targets ...)` — `Cytnx::cytnx` becomes a SHARED imported target; worth re-auditing which transitive deps (Boost, OpenMP, LAPACK) can move from PUBLIC to PRIVATE now that they're no longer forced PUBLIC by the "static archive can't absorb symbols" constraint.

**Re-validation, not new code, but real work:**
- Full rebuild + retest on `debug-openblas-cpu`/`debug-mkl-cpu`, and both the build-tree and install-tree consumer paths in `ci-downstream-find-package.yml`.
- `openblas-cuda`: `CUDA_SEPARABLE_COMPILATION`/`CUDA_RESOLVE_DEVICE_SYMBOLS` are already set on the `cytnx` target; shared vs. static device-code relocatable linking differs enough to need actual GPU hardware to confirm — not verifiable in a GPU-less sandbox.
- macOS: `@rpath`/install-name handling for the new `libcytnx.dylib`, analogous to what `pycytnx` already has, needs setting up and testing on real macOS CI.

## What this buys

- The LTO/GIMPLE-IR bloat problem from #947 disappears as a side effect for this artifact — a properly-linked `.so` holds real machine code, not raw archived LTO IR.
- Restores `cytnx.__cpp_include__`/`__cpp_lib__` and the documented "build custom C++ against your installed `pip`/`conda` cytnx" workflow, in an ABI-correct way instead of either leaving it broken or reintroducing the bloated static archive.

Note this is *not* about further wheel-size reduction — that's already solved by #967's `install.components` fix regardless of this decision. The value here is specifically restoring the C++-interop feature correctly, if the project wants it back.

---
Investigated by Claude Code on behalf of @IvanaGyro.

**Comment by ianmccul** [comment]
Note that projects like PyTorch probably have around 5 - 15 full time developers maintaining the C++ extension API boundary, and that isn't stable either: you need to recompile the extensions on every torch version, CUDA version, C++ ABI change, etc.  For Cytnx it is probably more than one full time job just keeping track of upstream changes, making decisions on whether new or changed functions become part of the exported API or not, adding tests, versioning support, etc.  All to save installing from CMake.  

- choose which headers are public and stable enough to ship;
- add export macros for every public class/function that downstream code needs;
- decide whether ABI stability is promised across patch/minor releases;
- define which compiler and C++ standard combinations are supported;
- define how Boost, BLAS/LAPACK, HPTT, CUDA, cuTENSOR, and cuQuantum appear to downstream builds;
- decide whether a Python wheel may expose CPU-only, CUDA, or multiple backend variants;
- test downstream extension builds in CI on Linux and macOS;
- manage RPATH/install-name behavior for the Python package layout;
- avoid shipping a huge static archive or LTO IR payload in the wheel;
- document a stable extension build API rather than asking users to scrape flags.

The only reason to do this is if there was an existing, thriving ecosystem of C++ developers who are writing extensions for Cytnx and want to distribute them to users without those users needing to compile Cytnx from source.  From searching Github for publically available extension projects intended to use a precompiled Cytnx shared library, I could find exactly two: 

### 1. `j9263178/CytnxBenchmarks`

Repository:

- <https://github.com/j9263178/CytnxBenchmarks>
- description: "Benchmarks for the Cytnx library."
- created: 2022-02-28
- last pushed: 2023-08-18
- latest commit seen: `edf095af`, message `fix svd dmrg`

Relevant file:

- <https://github.com/j9263178/CytnxBenchmarks/blob/edf095afc4bce5b5db100e6d714206d7747067eb/Makefile>

This is a real use of the old Python introspection mechanism. The Makefile actively computes:

- `CYTNX_INC` from `cytnx.__cpp_include__`
- `CYTNX_LINK` from `cytnx.__cpp_linkflags__`
- `CYTNX_LIB` from `cytnx.__cpp_lib__` plus `/libcytnx.a`
- `CYTNX_CXXFLAGS` from `cytnx.__cpp_flags__`

It then uses those values in a `g++` command to build benchmark executables. The repository appears to benchmark Cytnx against ITensor and Google Benchmark, with targets such as `dmrg_dense_test.e`.

Assessment:

- This is the strongest public evidence for the mechanism being used.
- It is not a reusable extension package or application framework; it is a local benchmark harness.
- It has hard-coded local paths such as `/home/petjelinux/Cytnx_lib`, `/home/petjelinux/itensor`, and `/home/petjelinux/benchmark`.
- It has not been pushed since August 2023.

So this is real, but it is old and looks like local research/development infrastructure rather than a maintained downstream package depending on a stable Cytnx C++ ABI.

### 2. `victorwk93812/Cytnx-fork-projects`

Repository:

- <https://github.com/victorwk93812/Cytnx-fork-projects>
- description: "Private Cytnx fork and Cytnx related/testing projects."
- created: 2025-11-09
- last pushed: 2026-02-24
- latest commit seen: `853ff436`, message `Cytnx dockerfile cpu in progress`

Relevant files:

- <https://github.com/victorwk93812/Cytnx-fork-projects/blob/853ff436ac8a35776b719128bef815b01cc10ad4/projects/cytnx-cpp-test/Makefile>
- <https://github.com/victorwk93812/Cytnx-fork-projects/blob/853ff436ac8a35776b719128bef815b01cc10ad4/projects/cytnx-cpp-test/cytnx-test.cpp>

The Makefile actively uses the same mechanism:

- `CYTNX_INC` from `cytnx.__cpp_include__`
- `CYTNX_LDFLAGS` from `cytnx.__cpp_linkflags__`
- `CYTNX_LIB` from `cytnx.__cpp_lib__` plus `/libcytnx.a`
- `CYTNX_CXXFLAGS` from `cytnx.__cpp_flags__`

The C++ file is a minimal smoke test:

```cpp
#include <iostream>
#include "cytnx.hpp"
using namespace std;
namespace cy=cytnx;
int main(){
    cout << "Hello World!" << endl;
    auto A = cy::Tensor({2, 3});
    auto B = A;

    cout << cy::is(B, A) << endl;
    auto C = cy::ones(4);
    cout << C << endl;
    return 0;
}
```
Assessment:

- This is a real use of the documented mechanism.
- It is essentially a compile/link smoke test.
- The repository README describes the repo as a private fork plus related/testing projects, not as an external library/application.
- It is useful evidence that someone copied and tried the documented workflow, but weak evidence for a mature downstream extension use case.

## What C++ Extensions Might Want To Add

You also need to consider what a Cytnx C++ API extension mechanism would want to add.  The existing C++ API is very Python-like and does not really add anything to what you can do in Python anyway.  There is little, if any, speed advantage from using the C++ API because it still runs through a dynamic dispatch mechanism, and in some cases the naive C++ version is slower than the equivalent Python code.

The sorts of things that you might want to do with a C++ extension could be
- interoperability with another code such as iTensor or an in-house tensor network code
- Eigen/Armadillo/Blaze-style dense linear algebra helpers
- HDF5 or domain-specific binary I/O layers
- Cusomized `LinOp`-style operators that are not easily expressed in Python
- Cusom CUDA kernels
- Alternative backends such as ROCm

These would be hard enough with the existing C++ source, but designing a stable extension API with these uses in mind is a huge amount of work, for a mechanism that has had **two** public users in the last 4 years, and both of them are rather minimal examples, a long way from a production extension that could be distributed to users. And I think they are both current or former NTU students who are quite capable of installing Cytnx from CMake.
